MQTT improvements and fixes

rechrtb opened the pull request:
Follow up to https://github.com/Duet3D/RepRapFirmware/pull/589.

- Change subscription max QOS parameter from `Q` to `O` in `M586.4`
- Removed single default topic configuration in `M586.4`. 
- Topic to publish to must be specified in `M118` via parameter `T` . QOS, retain and duplicate flags can be optionally specified via `Q`, `R` and `D` in `M118`
- Support LWT (last will and testament) QOS and retain flag via parameter `Q` and `R` in `M586.4`.
- MQTT demonstration script and documentation moved to [Duet3D/MQTT-WPA2-Enterprise-Demo](https://github.com/rechrtb/Duet-Demo).

Closes https://github.com/Duet3D/RepRapFirmware/issues/880.

rechrtb committed:
Support specifying will qos and retain

rechrtb committed:
Use unique arg for subscribe max qos

rechrtb committed:
Implement setting topic, qos, retain and dup on M118 MQTT publish

rechrtb committed:
Remove unneeded variable

rechrtb committed:
Remove need for intermediate output buffer when publishing

rechrtb committed:
Some cleanups

rechrtb committed:
Make publish topic mandatory

rechrtb committed:
Fix freeze when disabling MQTT

rechrtb committed:
Create only one mqtt client instance for now

rechrtb committed:
Set interface null properly

rechrtb committed:
Check if client can be enabled on multiple interfaces

rechrtb committed:
Remove mqtt demo files

This is moved to another repo

rechrtb committed:
Fixes for MQTT client

- Fix setting connect flags related to LWT
- Display debug message when changing subscription QOS
- Replace max QOS argument

rechrtb commented:
@PaulG4H, if you want to test early :) I'm testing with the setup from https://github.com/rechrtb/Duet-Demo/tree/master/mqtt, and seems to work ok when following the README, but I'm still double checking the code if I miss some potential bugs.

dc42 commented on src/GCodes/GCodes2.cpp:
```diff
@@ -2079,14 +2079,57 @@ bool GCodes::HandleMcode(GCodeBuffer& gb, const StringRef& reply) THROWS(GCodeEx
+						int32_t qos = 0;
+						if (gb.Seen('Q'))
+						{
+							qos = gb.GetIValue();
```
Please use GetLimitedUIValue to get a value between 0 and 2, then we can eliminate the test of the value of qos and the error message. We have very little free flash memory left on Duet 2 so we need to avoid using space for dedicated range tests wherever possible.

rechrtb replied:
Fixed. Didn't know about GetLimitedUIValue, thanks for the heads up! Actually used TryGetLimitedUIValue for the optional parameters.

dc42 commented on src/GCodes/GCodes2.cpp:
```diff
@@ -2079,14 +2079,57 @@ bool GCodes::HandleMcode(GCodeBuffer& gb, const StringRef& reply) THROWS(GCodeEx
+						int32_t retain = 0;
+						if (gb.Seen('R') && result != GCodeResult::error)
+						{
+							retain = gb.GetIValue();
```
As above please use GetLimitedUIValue.

dc42 commented on src/GCodes/GCodes2.cpp:
```diff
@@ -2079,14 +2079,57 @@ bool GCodes::HandleMcode(GCodeBuffer& gb, const StringRef& reply) THROWS(GCodeEx
+						if (gb.Seen('D') && result != GCodeResult::error)
+						{
+							dup = gb.GetIValue();
+							if (dup < 0 || dup > 1)
```
As above please use GetLimitedUIValue.

dc42 commented on src/Networking/MQTT/mqtt_pal.cpp:
```diff
@@ -53,4 +54,27 @@ ssize_t mqtt_pal_recvall(mqtt_pal_socket_handle fd, void* buf, size_t sz, int fl
+	char name[sizeof(format)] = { 0 };
+	SafeSnprintf(name, sizeof(name), format, count);
+	mutex->m = new Mutex();
+	mutex->m->Create(name);
```
This is not safe because the mutex will point to a name in memory that has been overwritten. As the names of mutexes are only used for debugging purposes and then only rarely, we could consider giving them all the same name.

rechrtb replied:
I had mistakenly thought that `Create` makes a copy of the string, modified to use the same name. As you said, it's only for debugging.

dc42 commented on src/Networking/MQTT/MqttClient.h:
```diff
@@ -55,32 +57,28 @@ class MqttClient : public NetworkClient
 	uint32_t messageTimer;	// General purpose variable for keeping track of queued messages timeout
-	Mutex publishMutex;
 
 	MqttClient *next;
```
Under what conditions can we have more than one MqttClient?

rechrtb replied:
The case where more than one MqttClient instance is needed is for when there are multiple brokers - either on the same interface or on different interfaces. 

Per our email conversation, only one broker on one interface needs to be supported for now. So I changed MqttClient to be a singleton. This is remnant of the old code I forgot to delete.

dc42 commented on src/Networking/MQTT/mqtt_pal.h:
```diff
@@ -26,11 +26,14 @@ typedef time_t mqtt_pal_time_t;
-#define MQTT_PAL_MUTEX_LOCK(x)
-#define MQTT_PAL_MUTEX_INIT(x)
-#define MQTT_PAL_MUTEX_UNLOCK(x)
+typedef struct
```
Please confirm that we need to use a mutex, i.e. there can be more than one task accessing those parts of the system that the mutex guards. Also, are we certain that no deadlock can occur?

rechrtb replied:
The mutex is for guarding resources between (1) the main task when it calls `mqtt_publish` - when buffering the message for sending and (2) the network task when it calls `mqtt_sync` - when moving data from mqtt buffer to socket buffer.

As to whether deadlocks can occur,  one of the condition for it does not seem to be satisfied - circular wait. `mqtt_sync` and `mqtt_publish` only share 1 common mutex they are both attempting to lock; and `mqtt_publish` itself only tries to attempt to lock 1 mutex, breaking the possibility of a circular wait chain from forming.

rechrtb committed:
Use GetLimitedUIValue for getting qos, retain and dup values

rechrtb committed:
Use same names for mqtt client mutex

rechrtb committed:
Remove unecessary code for multiple mqtt clients

Since MQTT client is implemented as a singleton for now, remove code
related to partially handling multiple instances.

dc42 commented:
Further to my previous message, I see that the current API does support
starting at a specified channel when connecting to a specified SSID. See
https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/network/esp_wifi.html#_CPPv417wifi_sta_config_t
.

On Mon, 25 Sept 2023 at 08:39, rechrtb ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/Networking/MQTT/MqttClient.h
> <https://github.com/rechrtb/RepRapFirmware/pull/2#discussion_r1335500756>:
>
> >
>  	uint32_t messageTimer;	// General purpose variable for keeping track of queued messages timeout
> -	Mutex publishMutex;
>
>  	MqttClient *next;
>
> The case where more than one MqttClient instance is needed is for when
> there are multiple brokers - either on the same interface or on different
> interfaces.
>
> Per our email conversation, only one broker on one interface needs to be
> supported for now. So I changed MqttClient to be a singleton. This is
> remnant of the old code I forgot to delete.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rechrtb/RepRapFirmware/pull/2#discussion_r1335500756>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAUYI3FIQUHMRNCHJJ3NC73X4EYLJANCNFSM6AAAAAA4QNCYPQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
David Crocker, Duet3D Ltd.

rechrtb committed:
Clear send buffer error manually

PaulG4H commented:
@rechrtb sorry for the late reply, 

Many thank's for your work to bring mqtt native to RRF!

I testet the current 3.5.0-rc.1+102 Version today, it works better but the Topic in M118 gets corrupted --> when I run:
M118 P6 S"idle" T"devices/rrf-k40/status" R1

The broker receives only:
![image](https://github.com/rechrtb/RepRapFirmware/assets/8039580/00ad12f5-71ea-4a58-b34e-e3ce3eb48afc)

I have no clue why the topic is corrupted to only >k40<, instead of the full path >devices/rrf-k40/status<

After further testing, the topic is ignored totaly, instead the machine Name is used.

It seams that my ESP32 Firmware is not working correctly?

rechrtb commented:
Hi @PaulG4H, no worries. 

I tested sending `M118 P6 S"idle" T"devices/rrf-k40/status" R1` with the setup from [the demo](https://github.com/Duet3D/MQTT-WPA2-Enterprise-Demo/tree/master/mqtt), and seems to work ok. No topic corruption or anything like that.

I will have to ask where did you get the binary for the RRF version you tested? You might be testing one that don't have these changes yet, or at least the latest changes in this pull request. Can you try doing `M118` without the `T` parameter? If it doesn't complain of a missing parameter, it means you're on an older version without these changes.

I'm not sure if you are able to build RRF with the changes in this pull request. If not, I can send you a build - though don't know the  board you're using. Furthermore, you might be apprehensive of trusting some binary from some stranger online (rightfully so!).

@dc42 can I share with @PaulG4H a built binary with this PR using the Duet3D Dropbox?

PaulG4H commented:
@rechrtb I already post this issue in the RRF discord --> I use a SKR2 board with the binary created by GloomyAndy and his Version is not the latest.

So I will wait for that.

I also see in the official documentation the LWT config, this is so great either!

When your changes are applied to the RRF binaries this opens a whole now world!

Smarthome Integration without pooling, no matter if you use Home Assistant, IP Symcon, IO Broker, OpenHab or any other which is able to receive MQTT topics!

Or switch a external relay from GCode (Tasmota or Shelly Device) directly over WiFi for example the Dust Collector for the CNC or the Chiller for the Laser...

Many thanks for your time and effort!

rechrtb commented:
Unfortunately I don't have the build environment for RRF SKR setup. So if
you want to test these changes, you might want to ask around if someone can
do a build for you with the branch on this PR. Or, wait for this to get
merged and RRF SKR project to pick up the changes like you mentioned.

Great to hear about your excitement regarding this feature! Much
appreciated.





On Sun, Oct 1, 2023, 18:31 PaulG4H ***@***.***> wrote:

> @rechrtb <https://github.com/rechrtb> I already post this issue in the
> RRF discord --> I use a SKR2 board with the binary created by GloomyAndy
> and his Version is not the latest.
>
> So I will wait for that.
>
> You also see in the official documentation the LWT config, this is so
> great!
>
> When your changes are applied to the RRF binaries this opens a whole now
> world!
>
> Smarthome Integration without pooling, no matter if you use Home
> Assistant, IP Symcon, IO Broker, OpenHab or any other which is able to
> receive MQTT topics!
>
> Or switch a external relay from GCode (Tasmota or Shelly Device) directly
> over WiFi for example the Dust Collector for the CNC or the Chiller for the
> Laser...
>
> Many thanks for your time and effort!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rechrtb/RepRapFirmware/pull/2#issuecomment-1742021298>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AYN7X3RG4LDJHXJ5MFSDBPLX5EZ6FANCNFSM6AAAAAA4QNCYPQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

dc42 commented:
@rechrtb  please make that change and then merge this PR into your 3.5-dev branch; then let me know and I will merge your 3.5-dev branch into the master.

rechrtb commented:
Closing due to mistake in target branch. See https://github.com/Duet3D/RepRapFirmware/pull/923 instead.
